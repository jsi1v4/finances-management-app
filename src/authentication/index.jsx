import React, { createContext, useState, useContext, useEffect } from 'react';
import { useRouter } from 'next/router';

export const AuthenticationContext = createContext(null);

export function AuthenticationProvider({ children }) {
  const [userName] = useState('John Doe');
  const [isAuth, setIsAuth] = useState(false);

  const authOn = () => {
    setIsAuth(true);
  };

  const authOff = () => {
    setIsAuth(false);
  };

  return (
    <AuthenticationContext.Provider
      value={{
        isAuth,
        userName,
        authOn,
        authOff
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}

export function useLoginAction() {
  const router = useRouter();
  return () => router.push('/login');
}

export function withAuth() {
  return (Component) =>
    React.memo(function AuthWrappedComponent(props) {
      const { isAuth } = useContext(AuthenticationContext);
      const loginAction = useLoginAction();

      useEffect(() => {
        if (!isAuth) {
          loginAction();
        }
      }, [isAuth, loginAction]);

      if (isAuth) {
        return <Component {...props} />;
      }
      return null;
    });
}
