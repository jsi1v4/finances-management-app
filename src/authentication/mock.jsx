import React from 'react';

import { AuthenticationContext } from './index';

export function AuthenticationMock({ children }) {
  return (
    <AuthenticationContext.Provider
      value={{
        isAuth: true,
        userName: 'John Doe',
        authOn: jest.fn(),
        authOff: jest.fn()
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}

export default AuthenticationMock;
