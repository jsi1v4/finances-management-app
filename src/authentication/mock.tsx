import React from 'react';

import { AuthenticationContext } from './index';

export function AuthenticationMock({ children }) {
  return (
    <AuthenticationContext.Provider
      value={{
        isAuth: true,
        session: null,
        authOn: jest.fn(),
        authOff: jest.fn()
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}

export default AuthenticationMock;
