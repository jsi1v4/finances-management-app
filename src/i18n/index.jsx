import React, { createContext, useState } from 'react';
import { useIntl } from 'react-intl';

import { INIT_LOCALE } from '../config';

export function useI18n() {
  const intl = useIntl();
  return (id, values) => {
    const message = intl.formatMessage({ id }, values);
    return <>{message}</>;
  };
}

export const I18nContext = createContext(null);

export function I18nProvider({ children }) {
  const [lang, setLanguage] = useState(INIT_LOCALE);
  const changeLang = (value) => {
    setLanguage(value);
  };

  return (
    <I18nContext.Provider
      value={{
        lang,
        changeLang
      }}
    >
      {children}
    </I18nContext.Provider>
  );
}

export default I18nProvider;
