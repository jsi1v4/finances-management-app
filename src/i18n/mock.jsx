import React from 'react';

import { I18nProvider } from './index';
import { I18nMessagesProvider } from './messages';

export function I18nMock({ children }) {
  return (
    <I18nProvider>
      <I18nMessagesProvider>{children}</I18nMessagesProvider>
    </I18nProvider>
  );
}

export default I18nMock;
