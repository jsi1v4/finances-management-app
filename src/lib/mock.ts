export class Auth {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  onAuthStateChanged(fn = () => {}) {
    return () => {};
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async setPersistence(persist?: string) {
    return new Promise((resolve) => resolve({}));
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async signInWithEmailAndPassword(username: string, password: string) {
    return new Promise((resolve) => resolve({}));
  }

  async signOut() {
    return new Promise((resolve) => resolve({}));
  }
}

export default Auth;
