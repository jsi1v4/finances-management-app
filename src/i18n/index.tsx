import React, { createContext, useContext, useState } from 'react';
import { useIntl } from 'react-intl';

import { INIT_LOCALE } from 'src/config';
import { II18nContext, Locale } from './types';

export const I18nContext = createContext<II18nContext>(null as never);

export const useI18n = () => useContext(I18nContext);

export function I18nProvider({ children }) {
  const [lang, setLanguage] = useState<Locale>(INIT_LOCALE);
  const changeLang = (value: Locale) => {
    setLanguage(value);
  };

  return (
    <I18nContext.Provider
      value={{
        lang,
        changeLang
      }}
    >
      {children}
    </I18nContext.Provider>
  );
}

export function useI18nMessage() {
  const intl = useIntl();
  return (id: string, values?: Record<string, string>) => {
    const message = intl.formatMessage({ id }, values);
    return message;
  };
}
