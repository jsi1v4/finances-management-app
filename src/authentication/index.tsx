import React, {
  PropsWithChildren,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState
} from 'react';

import { AuthInstance, SessionUser } from 'src/lib';
import { AuthService } from './service';
import { LoginProps, Session, IAuthenticationContext } from './types';

interface ProvidersProps {
  api: AuthInstance;
}

export const AuthenticationContext = createContext<IAuthenticationContext>(
  null as never
);

export const useAuth = () => useContext(AuthenticationContext);

export function AuthenticationProvider({
  api,
  children
}: PropsWithChildren<ProvidersProps>) {
  const service = useMemo(() => new AuthService(api), [api]);

  const [isAuth, setIsAuth] = useState<boolean | null>(null);
  const [session, setSession] = useState<Session | null>(null);

  const authOn = useCallback(
    async (props: LoginProps) => {
      await service.setPersistence(props.remember);
      await service.signIn(props.username, props.password);
    },
    [service]
  );

  const authOff = useCallback(async () => {
    await service.signOut();
  }, [service]);

  const onStateChanged = async (state: SessionUser) => {
    setIsAuth(!!state);
    if (state) {
      setSession({
        displayName: state.displayName,
        email: state.email,
        photoURL: state.photoURL,
        emailVerified: state.emailVerified,
        token: await state.getIdToken()
      });
    }
  };

  useEffect(() => {
    const unsubscribe = api.onAuthStateChanged(onStateChanged);
    return () => unsubscribe();
  }, [api]);

  return (
    <AuthenticationContext.Provider
      value={{
        isAuth,
        session,
        authOn,
        authOff
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}

export function useLoginAction() {
  return () => window.location.replace('/login');
}

export function withAuth() {
  return (Component) =>
    React.memo((props) => {
      const { isAuth } = useAuth();
      const loginAction = useLoginAction();

      useEffect(() => {
        if (isAuth === false) {
          loginAction();
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
      }, [isAuth]);

      if (isAuth) {
        return <Component {...props} />;
      }
      return null;
    });
}
