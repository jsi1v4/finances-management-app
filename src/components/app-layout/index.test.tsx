import React from 'react';
import { render } from '@testing-library/react';

import { wrapse } from 'src/utils';
import { I18nMock } from 'src/i18n/mock';
import { AuthenticationMock } from 'src/providers/authentication/mock';
import { AppLayout } from './index';

describe('AppLayout component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<AppLayout />, {
      wrapper: wrapse(I18nMock, AuthenticationMock)
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render component', async () => {
    const { getByTestId, getByText } = render(<AppLayout />, {
      wrapper: wrapse(I18nMock, AuthenticationMock)
    });
    expect(getByTestId('app-layout').innerHTML).toBeDefined();
    expect(getByText('Finances Management').innerHTML).toBeDefined();
  });
});
