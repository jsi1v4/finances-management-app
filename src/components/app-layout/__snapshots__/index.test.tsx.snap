// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppLayout component matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="ant-layout ant-layout-has-sider"
    data-testid="app-layout"
  >
    <aside
      class="ant-layout-sider ant-layout-sider-dark"
      style="flex: 0 0 200px; max-width: 200px; min-width: 200px; width: 200px;"
    >
      <div
        class="ant-layout-sider-children"
      >
        <div
          class="ant-row styles__Centered-sc-14bh8a3-4 eVRgfD"
        >
          <img
            class="styles__Logo-sc-14bh8a3-0 fnKFrM"
            src="https://raw.githubusercontent.com/jsi1v4/jsi1v4/main/assets/logo192.png"
          />
        </div>
      </div>
    </aside>
    <section
      class="ant-layout styles__FullPage-sc-14bh8a3-1 hPVqfu"
    >
      <div
        class="ant-page-header styles__Header-sc-14bh8a3-2 kYvdZv ant-page-header-ghost"
      >
        <div
          class="ant-page-header-heading"
        >
          <div
            class="ant-page-header-heading-left"
          >
            <span
              class="ant-page-header-heading-title"
              title="My planning"
            >
              My planning
            </span>
          </div>
          <span
            class="ant-page-header-heading-extra"
          />
        </div>
      </div>
      <section
        class="ant-layout styles__Content-sc-14bh8a3-3 hHHrMw"
      />
    </section>
  </section>
</DocumentFragment>
`;
